### PR TITLE
Allow user to specify length of prompt

### PR DESCRIPTION
**Pull request by AlexWertheim**
I've added an argument called `prompt-len` which the user can specify via the command line. The prompt is now set to be the sentence composed of the first `prompt-len` tokens in order. Note that there is currently no error checking; for example, the program does not check whether `prompt-len` exceeds the vocab size.

### CODE REVIEW

File: example_xla.py [hunk 1]
```diff
@@ -96,9 +97,11 @@ def main(
         ckpt_dir, tokenizer_path, rank, world_size, max_seq_len, max_batch_size, dim, n_layers, n_heads
     )
 
-    prompts = [
+    prompts = [generator.tokenizer.decode(list(range(prompt_len)))]
+    print(prompts)
```
Comment (JackCaoG) [inline example_xla.py:101]: is this for debug only?

Comment (AlexWertheim) [reply example_xla.py:101]: Yeah, I'll remove the print statement once all other comments are settled.

File: example_xla.py [hunk 2]
```diff
@@ -159,11 +163,12 @@ def mp_main(
     dim: int = 4096,
     n_layers: int = 32,
     n_heads: int = 32,
+    prompt_len: int = 6,
```
[the remark this answers is not in the document]
Comment (AlexWertheim) [reply example_xla.py:166]: I set the default to 6 because I wanted the default to be the same number of input tokens as our previous prompt "I believe the meaning of life is", only I miscounted (that has 7 words), and there's not a 1-1 mapping between words and tokens necessarily, so it's still wrong regardless. We can change the default if the current one is not right - any suggestions on alternatives?

If I'm reading the code right, `max_batch_size` is related to the total number of prompts, not the number of tokens in each prompt. The goal of the `prompt_len` is to allow the user to specify a variable number of input tokens in a single prompt. (It is reasonable to point out that this does not currently support multiple prompts.) I'm not sure I understand your comments about `bs = 1` or about whether this is the right solution. (Whether what is the right solution?) Could you please clarify?

[the remark this answers is not in the document]
Comment (alanwaketan) [reply example_xla.py:166]: Okay, now I get it. But from the discussion in gchat, do you still need this approach?

[the remark this answers is not in the document]
Comment (AlexWertheim) [reply example_xla.py:166]: If you mean our discussion about `max_seq_len` and `max_gen_len`, this is something quite separate, right? This PR allows the user to modify the length of the input prompt. `max_seq_len` controls the size allocated for the output (and in our repo, the total number of tokens generated), and `max_gen_len` controls the number of tokens displayed. I think we still need this for the user to modify input - please let me know if you had some other discussion in mind.

**Review by alanwaketan — APPROVED** [review]
LGTM. Please update the user guide and the README in this branch.

**Comment by Liyang90** [comment]
LGTM

**Comment by JackCaoG** [comment]
@AlexWertheim do you want to just merge this pr or you want to put it on hold for now

**Comment by AlexWertheim** [comment]
> @AlexWertheim do you want to just merge this pr or you want to put it on hold for now

I'm fine either way. I think it'd be best to merge the changes into the stable branch, especially with Liyang's improvements now merged on top of mine. I think @miladm wanted to give a review though, so I was waiting on his feedback to merge.

File: llama/generation.py [hunk 3]
```diff
@@ -58,33 +58,57 @@ def generate(
...
-        output_pos_tensor = cur_pos_tensor - 1
-        input_tokens = tokens.index_select(1, input_pos_tensor)
         cache_kvs = self.model.cache_kvs
-        xm.mark_step(wait=True)
+        xm.mark_step()
 
         decoding_start_time = time.time()
-        for _ in range(start_pos, total_len):
+        prev_pos = 0
+        scale_factor = 8
```
Comment (miladm) [inline llama/generation.py:79]: please add a comment to explain why we hard code this value to `8`.

File: llama/generation.py [hunk 4]
```diff
@@ -58,33 +58,57 @@ def generate(
...
+        while prev_pos < min_prompt_size:
```
Comment (miladm) [inline llama/generation.py:80]: Please add a comment with a couple of sentences explaining this bucketization scheme. I expect this to be a useful example use case for future OSS dev work.

File: llama/generation.py [hunk 5]
```diff
@@ -58,33 +58,57 @@ def generate(
...
+            section_len = 1
+            while prev_pos + section_len * scale_factor <= min_prompt_size:
+                section_len *= scale_factor
+            cur_pos = prev_pos + section_len
+            print(f"Processing prompt pos [{prev_pos}, {cur_pos}), section length {section_len}")
+            cur_pos_tensor = torch.tensor(cur_pos).to(device)
+            input_pos_tensor = torch.arange(prev_pos, cur_pos).to(device)
+            output_pos_tensor = cur_pos_tensor - 1
+            input_tokens = tokens.index_select(1, input_pos_tensor)
+            xm.mark_step()
```
Comment (miladm) [inline llama/generation.py:90]: why the. additional mark_step here?

**Review by miladm — COMMENTED** [review]
Thanks. Can we add a few words on the details of our measurement methodology before and after this change?

File: llama/generation.py [hunk 6]
```diff
@@ -58,33 +58,57 @@ def generate(
 
         prompt_tokens = [self.tokenizer.encode(x, bos=True, eos=False) for x in prompts]
 
-        total_len = params.max_seq_len
+        min_prompt_size = min([len(t) for t in prompt_tokens])
+        max_prompt_size = max([len(t) for t in prompt_tokens])
+        assert min_prompt_size >= 1 and max_prompt_size < params.max_seq_len
```
Comment (miladm) [inline llama/generation.py:63]: shouldn't this be `max_gen_len + max_prompt_size < params.max_seq_len`?

File: example_xla.py [hunk 7]
```diff
@@ -139,31 +142,35 @@ def _fn(
...
-    main(tokenizer_path, temperature, top_p, max_seq_len, max_batch_size, ckpt_dir, dim, n_layers, n_heads)
+    main(tokenizer_path, temperature, top_p, max_seq_len, max_batch_size, ckpt_dir, dim, n_layers, n_heads, prompt_len, max_gen_len)
 
 def mp_main(
     mp: bool,
     tokenizer_path: str,
     temperature: float = 0.8,
     top_p: float = 0.95,
-    max_seq_len: int = 512,
+    max_seq_len: int = 2048,
```
Comment (miladm) [inline example_xla.py:161]: can we add a comment to define `max_seq_len`, `prompt_len`, `max_gen_len` to clarify for the user in plain English?

File: llama/generation.py (hunk 6 above)
Comment (Liyang90) [reply llama/generation.py:63]: `max_gen_len + max_prompt_size` can be larger than `max_seq_len`, and in that case, the actual generation length would be truncated to fit in `max_seq_len`.
`max_prompt_size < params.max_seq_len` is there to make sure there is at least 1 token that can be generated for the longest prompt.

File: llama/generation.py (hunk 5 above)
Comment (Liyang90) [reply llama/generation.py:90]: This is to cut the graph so the call to the function below creates a clean graph that can be reused.

File: llama/generation.py (hunk 4 above)
Comment (Liyang90) [reply llama/generation.py:80]: This is a little too much to explain in a comment. To let user understand this, we either point to the PR https://github.com/pytorch-tpu/llama/pull/15 or I can make a doc for future reference.

File: llama/generation.py (hunk 3 above)
Comment (Liyang90) [reply llama/generation.py:79]: This is a little too much to explain in a comment. To let user understand this, we either point to the PR https://github.com/pytorch-tpu/llama/pull/15 or I can make a doc for future reference.
